File: server/python/src/demo/controllers/dataset.py
"""
Dataset Controller

使用依赖注入模式，session 通过 FastAPI Depends 注入。
"""

from fastapi import APIRouter, Depends, HTTPException
from fastapi.responses import ORJSONResponse
from sqlalchemy.ext.asyncio import AsyncSession

from demo.schemas.dataset import DatasetCreate, DatasetResponse, DatasetUpdate
from demo.services.dataset import dataset_service
from framework.database.dependencies import get_db_session
from framework.schemas.base import Success, SuccessExtra

router = APIRouter()


@router.get("")
async def list_datasets(
    page: int = 1,
    page_size: int = 10,
    session: AsyncSession = Depends(get_db_session),
) -> ORJSONResponse:
    """获取知识库列表"""
    total, datasets = await dataset_service.list_datasets(session, page, page_size)
    return SuccessExtra(
        data=[DatasetResponse.model_validate(d).model_dump() for d in datasets],
        total=total,
        page=page,
        page_size=page_size,
    )


@router.post("")
async def create_dataset(
    data: DatasetCreate,
    session: AsyncSession = Depends(get_db_session),
) -> ORJSONResponse:
    """创建知识库"""
    dataset = await dataset_service.create_dataset(session, data)
    return Success(data=DatasetResponse.model_validate(dataset).model_dump())


@router.get("/{dataset_id}")
async def get_dataset(
    dataset_id: str,
    session: AsyncSession = Depends(get_db_session),
) -> ORJSONResponse:
    """获取知识库详情"""
    dataset = await dataset_service.get_dataset(session, dataset_id)
    if not dataset:
        raise HTTPException(status_code=404, detail="知识库不存在")
    return Success(data=DatasetResponse.model_validate(dataset).model_dump())


@router.put("/{dataset_id}")
async def update_dataset(
    dataset_id: str,
    data: DatasetUpdate,
    session: AsyncSession = Depends(get_db_session),
) -> ORJSONResponse:
    """更新知识库"""
    dataset = await dataset_service.update_dataset(session, dataset_id, data)
    if not dataset:
        raise HTTPException(status_code=404, detail="知识库不存在")
    return Success(data=DatasetResponse.model_validate(dataset).model_dump())


@router.delete("/{dataset_id}")
async def delete_dataset(
    dataset_id: str,
    session: AsyncSession = Depends(get_db_session),
) -> ORJSONResponse:
    """删除知识库"""
    success = await dataset_service.delete_dataset(session, dataset_id)
    if not success:
        raise HTTPException(status_code=404, detail="知识库不存在")
    return Success(data=success)

File: server/rust/src/controllers/dataset.rs
//! Dataset 控制器

use axum::{
    extract::{Path, Query, State},
    Json,
};
use uuid::Uuid;
use validator::Validate;

use super::AppState;
use crate::common::error::{ApiResponse, Error, Result};
use crate::schemas::{DatasetCreate, DatasetUpdate, DatasetVo, PageRequest, PageResponse};

/// 获取知识库列表
#[utoipa::path(
    get,
    path = "/api/v1/datasets",
    params(PageRequest),
    responses(
        (status = 200, description = "获取成功", body = ApiResponse<PageResponse<DatasetVo>>)
    ),
    tag = "Dataset"
)]
pub async fn list_datasets(
    State(state): State<AppState>,
    Query(params): Query<PageRequest>,
) -> Result<Json<ApiResponse<PageResponse<DatasetVo>>>> {
    let result = state.dataset_service.list(params.page, params.page_size).await?;
    Ok(Json(ApiResponse::success(result)))
}

/// 创建知识库
#[utoipa::path(
    post,
    path = "/api/v1/datasets",
    request_body = DatasetCreate,
    responses(
        (status = 200, description = "创建成功", body = ApiResponse<DatasetVo>)
    ),
    tag = "Dataset"
)]
pub async fn create_dataset(
    State(state): State<AppState>,
    Json(data): Json<DatasetCreate>,
) -> Result<Json<ApiResponse<DatasetVo>>> {
    data.validate().map_err(|e| Error::Validation(e.to_string()))?;
    let result = state.dataset_service.create(data).await?;
    Ok(Json(ApiResponse::success(result)))
}

/// 获取知识库详情
#[utoipa::path(
    get,
    path = "/api/v1/datasets/{id}",
    params(
        ("id" = Uuid, Path, description = "知识库 ID")
    ),
    responses(
        (status = 200, description = "获取成功", body = ApiResponse<DatasetVo>),
        (status = 404, description = "知识库不存在")
    ),
    tag = "Dataset"
)]
pub async fn get_dataset(
    State(state): State<AppState>,
    Path(id): Path<Uuid>,
) -> Result<Json<ApiResponse<DatasetVo>>> {
    let result = state
        .dataset_service
        .get(id)
        .await?
        .ok_or_else(|| Error::NotFound("知识库不存在".to_string()))?;
    Ok(Json(ApiResponse::success(result)))
}

/// 更新知识库
#[utoipa::path(
    put,
    path = "/api/v1/datasets/{id}",
    params(
        ("id" = Uuid, Path, description = "知识库 ID")
    ),
    request_body = DatasetUpdate,
    responses(
        (status = 200, description = "更新成功", body = ApiResponse<DatasetVo>),
        (status = 404, description = "知识库不存在")
    ),
    tag = "Dataset"
)]
pub async fn update_dataset(
    State(state): State<AppState>,
    Path(id): Path<Uuid>,
    Json(data): Json<DatasetUpdate>,
) -> Result<Json<ApiResponse<DatasetVo>>> {
    if let Some(ref name) = data.name {
        if name.is_empty() || name.len() > 255 {
            return Err(Error::Validation("名称长度必须在1-255之间".to_string()));
        }
    }

    let result = state
        .dataset_service
        .update(id, data)
        .await?
        .ok_or_else(|| Error::NotFound("知识库不存在".to_string()))?;
    Ok(Json(ApiResponse::success(result)))
}

/// 删除知识库
#[utoipa::path(
    delete,
    path = "/api/v1/datasets/{id}",
    params(
        ("id" = Uuid, Path, description = "知识库 ID")
    ),
    responses(
        (status = 200, description = "删除成功", body = ApiResponse<()>),
        (status = 404, description = "知识库不存在")
    ),
    tag = "Dataset"
)]
pub async fn delete_dataset(
    State(state): State<AppState>,
    Path(id): Path<Uuid>,
) -> Result<Json<ApiResponse<()>>> {
    let success = state.dataset_service.delete(id).await?;
    if !success {
        return Err(Error::NotFound("知识库不存在".to_string()));
    }
    Ok(Json(ApiResponse::success_empty()))
}
